eventsources/persist/event_persist.go

```diff
@@ -0,0 +1,138 @@
+package persist
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/argoproj/argo-events/common"
+	"github.com/argoproj/argo-events/pkg/apis/eventsource/v1alpha1"
+)
+
+type EventPersist interface {
+	Save(event *Event) error
+	Get(key string) (*Event, error)
+	IsEnabled() bool
+}
+
+type Event struct {
+	EventKey     string
+	EventPayload string
+}
+
+type ConfigMapPersist struct {
+	kubeClient       kubernetes.Interface
+	name             string
+	namespace        string
+	createIfNotExist bool
+	lock             sync.Mutex
+}
+
+func createConfigmap(client kubernetes.Interface, name, namespace string) (*v1.ConfigMap, error) {
+	cm := v1.ConfigMap{}
+	cm.Name = name
+	cm.Namespace = namespace
+	return client.CoreV1().ConfigMaps(namespace).Create(&cm)
+}
+
+func NewConfigMapPersist(client kubernetes.Interface, configmap *v1alpha1.ConfigMapPersistence, namespace string) (EventPersist, error) {
+	if configmap == nil {
+		return nil, fmt.Errorf("persistence configuration is nil")
+	}
+	_, err := client.CoreV1().ConfigMaps(namespace).Get(configmap.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierr.IsNotFound(err) && configmap.CreateIfNotExist {
+			_, err = createConfigmap(client, configmap.Name, namespace)
+			if err != nil {
+				if !apierr.IsAlreadyExists(err) {
+					return nil, err
+				}
+			}
+		} else {
+			return nil, err
+		}
+	}
+	cmp := ConfigMapPersist{
+		kubeClient:       client,
+		name:             configmap.Name,
+		namespace:        namespace,
+		createIfNotExist: configmap.CreateIfNotExist,
+		lock:             sync.Mutex{},
+	}
+	return &cmp, nil
+}
+
+func (cmp *ConfigMapPersist) IsEnabled() bool {
+	return true
+}
+
+func (cmp *ConfigMapPersist) Save(event *Event) error {
+	if event == nil {
+		return errors.Errorf("event object is nil")
+	}
+	cmp.lock.Lock()
```

Comment:
I think the lock is useless, it does not resolved the race condition issue. Because:

1. If the ConfigMap is used by 2 EventSource objects, which means they run in different PODs, the lock does not help at all;
1. Even if the ConfigMap is referenced in same EventSource object twice (by 2 events config), they run in same POD, each of them will run a GoRoutine, and each of them has an `EventListener` instance, so each of them has a lock - does not help.

So maybe the best way we can do it just retry... thoughts?

Reply:
Yes, you are right. The lock is useless here. I thought multiple events are running on the same goroutine.  Retry and K8s server will handle the thread-safe. I will remove it.